### PR TITLE
Spring boot support

### PR DESCRIPTION
**Pull request by evenh**
This PR addresses #55 and provides an initial implementation for Spring Boot integration.

### CODE REVIEW

**Comment by kagkarlsson** [comment]
Awesome, thanks! Will have a look! :)

**Comment by evenh** [comment]
Please feel free to be frank in the review. I know Spring Boot better than the DB scheduler :-)

File: db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    @ConditionalOnBean(DataSource.class)
+    @ConditionalOnMissingBean
+    @Bean(initMethod = "start", destroyMethod = "stop")
+    public Scheduler scheduler(DbSchedulerCustomizer customizer) {
+        log.info("Creating DB Scheduler using tasks from Spring context: {}", configuredTasks);
+
+        if (existingDataSource instanceof TransactionAwareDataSourceProxy) {
+            log.info("Using a transaction aware DataSource");
+        } else {
+            log.info("The configured DataSource is not transaction aware: {}", existingDataSource);
```
Comment (kagkarlsson) [inline db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: Maybe we should say something about the consequences here, i.e. operations such as `schedule`, `reschedule`, `cancel` will have effect regardless of spanning transaction

File: db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/config/DbSchedulerProperties.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    public String getTableName() {
+        return tableName;
+    }
+
+    public void setTableName(final String tableName) {
+        this.tableName = tableName;
+    }
+
+    public boolean isImmediateExecution() {
```
Comment (kagkarlsson) [inline db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/config/DbSchedulerProperties.java]: Suffix with `isImmediateExecutionEnabled`?  (same for setter)

File: db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+@ConditionalOnBean(DataSource.class)
+public class DbSchedulerAutoConfiguration {
+    private static final Logger log = LoggerFactory.getLogger(DbSchedulerAutoConfiguration.class);
+
+    private final DbSchedulerProperties config;
+    private final DataSource existingDataSource;
+    private final List<Task<?>> configuredTasks;
+
+    public DbSchedulerAutoConfiguration(DbSchedulerProperties dbSchedulerProperties,
+        DataSource dataSource, List<Task<?>> configuredTasks) {
```
Comment (kagkarlsson) [inline db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: Will this inject all instances of `Task` on the classpath?

**Review by kagkarlsson — COMMENTED** [review]
Looks good, though probably will need a second scan of things :)

Would be neat with some short example `Main` showing a minimal example, ala for example https://github.com/kagkarlsson/db-scheduler/blob/master/src/test/java/com/github/kagkarlsson/scheduler/example/CronMain.java

**Comment by kagkarlsson** [comment]
Hmm, I think we are missing the `startTasks(...)` aspect for kicking off initial execution for recurring tasks, see https://github.com/kagkarlsson/db-scheduler/blob/master/src/test/java/com/github/kagkarlsson/scheduler/example/TasksMain.java#L41

Tasks that also implement `OnStartup` should be added to the builder via `startTasks(...)` method

File: db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java (hunk 3 above)
Comment (evenh) [reply db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: It should inject all `Task`-instances in the Spring context, or an empty list if no `Task` beans is defined.

Should we throw an exception if the list of configured `Task` beans is empty or just let the scheduler idle away?

**Comment by evenh** [comment]
I've implemented the `startTasks` invocation in the builder, where I've filtered out tasks which do and do not implement `OnStartup`.

As for the examples, should they be separate "real" Spring Boot projects? I was thinking something like (Maven hierarchy):

```
| db-scheduler
| db-scheduler-boot-starter
| examples
|  --> spring-boot-example
|  --> vanilla-java-example?
|  --> other-example?
```

The benefit of this approach is breakage of the examples if the library/integrations change :-)

File: db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @ConditionalOnBean(DataSource.class)
+    @ConditionalOnMissingBean
+    @Bean(initMethod = "start", destroyMethod = "stop")
+    public Scheduler scheduler(DbSchedulerCustomizer customizer) {
+        log.info("Creating DB Scheduler using tasks from Spring context: {}", configuredTasks);
+
+        if (existingDataSource instanceof TransactionAwareDataSourceProxy) {
+            log.info("Using a transaction aware DataSource");
```
Comment (cash1981) [inline db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: Is this really necessary information to give? I guess the else block info makes more sense than this?

Comment (evenh) [reply db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: Probably not worthy of an `info` but a `debug` might be handy?

File: db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        } else {
+            log.info("The configured DataSource is not transaction aware: '{}'. Operations such as schedule, reschedule and cancel will have effect regardless of spanning transactions.", existingDataSource);
```
Comment (cash1981) [inline db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: This log statement is a bit confusing. What is the purpose of this?

Comment (kagkarlsson) [reply db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: Are there cases where a Spring-user might send in a datasource is not transaction-aware by mistake?

**Comment by cash1981** [comment]
Perhaps also update the documentation on how to add db-scheduler using Spring and this PR?

File: db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java (hunk 5 above)
Comment (evenh) [reply db-scheduler-boot-starter/src/main/java/com/github/kagkarlsson/scheduler/boot/autoconfigure/DbSchedulerAutoConfiguration.java]: @cash1981: I thought that information whether one uses an instance of `TransactionAwareDataSourceProxy` would be useful information, see the [docs](https://github.com/kagkarlsson/db-scheduler#things-to-note--gotchas).

However, I can't seem to get Spring Boot's configuration to provide me with a `TransactionAwareDataSourceProxy` out of the box (using JPA or JDBC starter, `@EnableTransactionManagement` on/off and using a local PostgreSQL db).

**Comment by evenh** [comment]
Absolutely @cash1981! As pointed out [in this comment](https://github.com/kagkarlsson/db-scheduler/pull/63#issuecomment-514825576), we should consider adding some examples and link to those as well.

**Comment by kagkarlsson** [comment]
@evenh Yeah for spring-boot I think it makes sense to have an examples-directory with a full project that is part of the build. We can link to it from the documentation. If necessary, code can be  copied to the markdown-doc. Maybe a separate markdown-page for usage in spring-boot which we link to. 

If you could create an example project @evenh , I can update the doc :)

**Comment by evenh** [comment]
Fixed @kagkarlsson :)

**Comment by kagkarlsson** [comment]
Ran the example and shutdown seem to be working as exepected @cash1981

**Comment by kagkarlsson** [comment]
I think we need to wrap the datasource in a `TransactionAwareDataSourceProxy` if it is not already

**Comment by kagkarlsson** [comment]
Enabled maven-dependency-plugin:

```[INFO] --- maven-dependency-plugin:2.10:analyze-only (default) @ db-scheduler-spring-boot-starter ---
[WARNING] Used undeclared dependencies found:
[WARNING]    org.springframework.boot:spring-boot-test:jar:2.1.6.RELEASE:test
[WARNING]    org.assertj:assertj-core:jar:3.11.1:test
[WARNING]    org.springframework.boot:spring-boot-autoconfigure:jar:2.1.6.RELEASE:compile
[WARNING]    org.springframework.boot:spring-boot-actuator:jar:2.1.6.RELEASE:compile
[WARNING]    org.springframework.boot:spring-boot:jar:2.1.6.RELEASE:compile
[WARNING]    junit:junit:jar:4.12:test
[WARNING]    org.slf4j:slf4j-api:jar:1.7.26:compile
[WARNING]    org.springframework:spring-context:jar:5.1.8.RELEASE:compile
[WARNING] Unused declared dependencies found:
[WARNING]    org.springframework.boot:spring-boot-configuration-processor:jar:2.1.6.RELEASE:compile
[WARNING]    org.springframework.boot:spring-boot-starter:jar:2.1.6.RELEASE:compile
[WARNING]    org.springframework.boot:spring-boot-autoconfigure-processor:jar:2.1.6.RELEASE:compile```

Will see if I can fix a couple..

**Comment by kagkarlsson** [comment]
Merging this PR. Further improvements can be suggested as separate issues.

**Comment by kagkarlsson** [comment]
🎉 This issue has been resolved in `v16.9.0` ([Release Notes](https://github.com/kagkarlsson/db-scheduler/releases/tag/v16.9.0))

**Pull request merged**